biz.aQute.repository/src/aQute/p2/provider/P2Impl.java

```diff
@@ -75,20 +75,11 @@ private URI normalize(URI base) throws Exception {
 	private Promise<List<Artifact>> getArtifacts(Set<URI> cycles, URI uri) {
 		if (!cycles.add(uri)) {
 			return promiseFactory.resolved(Collections.emptyList());
```

Comment:
This should go back to how it was (returning a failed promise). The failed promise is properly handled at https://github.com/bndtools/bnd/blob/ce27a28a42a4cc3e9ca4740417ed99c4d3c8ffd2/biz.aQute.repository/src/aQute/p2/provider/P2Impl.java#L128-L133 where a message may be logged and promise result converted to an empty list.

Reply:
But I do not want a message?

Reply:
Looking at P2 cycles seem to be common in complex repos

Reply:
So is a cycle a normal possibility and not an error condition?

Reply:
Yup, they nowadays composite repositories that point to a very large number of other repositories. So it is likely not uncommon. It would be extremely hard to control there are no cycles since they often just point to a versionless URI.

So it should not generate errors imho.

Reply:
OK, then your change makes sense.